File: RepoTransBench/target_projects/C/Python/CCareaga_heap_allocator/tests/original/test_heap.py
import pytest

# Constants that are likely defined in heap.h; adjust if different
BIN_COUNT = 8
HEAP_INIT_SIZE = 4096

class Node:
    def __init__(self, size=0, next=None, prev=None):
        self.size = size
        self.next = next
        self.prev = prev

class Bin:
    def __init__(self):
        self.head = None

def add_node(bin_obj, node):
    node.next = bin_obj.head
    if bin_obj.head:
        bin_obj.head.prev = node
    node.prev = None
    bin_obj.head = node

def remove_node(bin_obj, node):
    curr = bin_obj.head
    prev = None
    while curr:
        if curr is node:
            if prev:
                prev.next = curr.next
            else:
                bin_obj.head = curr.next
            if curr.next:
                curr.next.prev = prev
            break
        prev = curr
        curr = curr.next

def get_best_fit(bin_obj, req_size):
    curr = bin_obj.head
    best = None
    best_size = None
    while curr:
        if curr.size >= req_size:
            if best is None or curr.size < best_size:
                best = curr
                best_size = curr.size
        curr = curr.next
    return best

class Heap:
    def __init__(self):
        self.bins = [Bin() for _ in range(BIN_COUNT)]
        self.region = None
        self.region_size = 0

def get_bin_index(size):
    # Simulated logic: find bin for size classes (e.g. 8, 16, 32, ...)
    # Here, just for test's sake, always return a valid index if size > 0
    if size <= 0:
        return -1
    return min(size.bit_length()-4, BIN_COUNT-1) if size >= 8 else 0

def init_heap(heap, region):
    heap.region = region
    heap.region_size = HEAP_INIT_SIZE

def heap_alloc(heap, size):
    # Simulated alloc: Just return an object if size reasonable, else None
    if size <= 0 or size > heap.region_size:
        return None
    # Simulate a growing region with a unique Python object
    return object()

def heap_free(heap, ptr):
    # Simulated free: does nothing, but would mark memory as free in real impl
    return

def expand(heap, sz):
    # Simulate always success
    return 0

def contract(heap, sz):
    # Simulate no-op
    return

def setup_heap(heap, region):
    # 'heap' is a Heap instance
    for i in range(BIN_COUNT):
        heap.bins[i] = Bin()
    init_heap(heap, region)

def cleanup_heap(heap):
    # No cleanup needed for this simulated Python version
    pass

def test_basic_alloc_free():
    heap = Heap()
    region = object()
    setup_heap(heap, region)

    p1 = heap_alloc(heap, 8)
    assert p1 is not None
    p2 = heap_alloc(heap, 32)
    assert p2 is not None

    heap_free(heap, p1)
    heap_free(heap, p2)

    cleanup_heap(heap)

def test_overflow_allocation():
    heap = Heap()
    region = object()
    setup_heap(heap, region)

    max_alloc = heap_alloc(heap, HEAP_INIT_SIZE)
    assert max_alloc is None

    cleanup_heap(heap)

def test_bin_index():
    assert get_bin_index(4) >= 0
    assert get_bin_index(64)  >= 0
    assert get_bin_index(1024) >= 0

def test_add_and_remove_node():
    bin_obj = Bin()
    n1 = Node(16)
    n2 = Node(32)
    n3 = Node(8)

    add_node(bin_obj, n1)
    assert bin_obj.head == n1

    add_node(bin_obj, n2)
    add_node(bin_obj, n3)

    assert bin_obj.head == n3
    assert bin_obj.head.next == n2
    assert bin_obj.head.next.next == n1

    remove_node(bin_obj, n2)
    assert bin_obj.head.next == n1

    remove_node(bin_obj, n3)
    assert bin_obj.head == n1

    remove_node(bin_obj, n1)
    assert bin_obj.head is None

def test_get_best_fit():
    bin_obj = Bin()
    n1 = Node(8)
    n2 = Node(32)
    n3 = Node(16)
    add_node(bin_obj, n2)
    add_node(bin_obj, n1)
    add_node(bin_obj, n3)

    fit = get_best_fit(bin_obj, 15)
    assert fit is not None and fit.size >= 15

    fit = get_best_fit(bin_obj, 100)
    assert fit is None

def test_split_and_merge():
    heap = Heap()
    region = object()
    setup_heap(heap, region)

    a = heap_alloc(heap, 8)
    b = heap_alloc(heap, 128)
    assert a and b
    heap_free(heap, b)
    c = heap_alloc(heap, 64)
    assert c

    heap_free(heap, a)
    heap_free(heap, c)

    cleanup_heap(heap)

def test_contract_and_expand():
    heap = Heap()
    ret = expand(heap, 0x1000)
    assert ret == 0
    contract(heap, 0x1000)

def test_free_only_head():
    heap = Heap()
    region = object()
    setup_heap(heap, region)

    a = heap_alloc(heap, 8)
    heap_free(heap, a)  # Should take the shortcut for heap head

    cleanup_heap(heap)

def test_alloc_min_size():
    heap = Heap()
    region = object()
    setup_heap(heap, region)

    a = heap_alloc(heap, 1) # Should get minimum allocation
    assert a is not None
    heap_free(heap, a)

    cleanup_heap(heap)

File: RepoTransBench/target_projects/C/Rust/CCareaga_heap_allocator/tests/original/test_heap.rs
use ccareaga_heap_allocator::{
    Heap, Bin, Node, add_node, remove_node, get_best_fit, get_last_node,
    get_bin_index, init_heap, heap_alloc, heap_free, expand, contract, BIN_COUNT, HEAP_INIT_SIZE,
};

fn setup_heap(heap: &mut Heap) {
    // Zero heap and bins, mimic C allocation
    for i in 0..BIN_COUNT {
        heap.bins[i] = Some(Bin::default());
    }
    // Use address as in C, but just a stub u64 here
    init_heap(heap, 0x10000);
}

fn cleanup_heap(_heap: &mut Heap) {
    // Rust drops bins automatically
}

#[test]
fn test_basic_alloc_free() {
    let mut heap = Heap::default();
    setup_heap(&mut heap);

    let p1 = heap_alloc(&mut heap,8);
    assert!(p1.is_some());
    let p2 = heap_alloc(&mut heap,32);
    assert!(p2.is_some());

    heap_free(&mut heap, &p1);
    heap_free(&mut heap, &p2);

    cleanup_heap(&mut heap);
}

#[test]
fn test_overflow_allocation() {
    let mut heap = Heap::default();
    setup_heap(&mut heap);

    let max_alloc = heap_alloc(&mut heap, HEAP_INIT_SIZE);
    assert!(max_alloc.is_none());

    cleanup_heap(&mut heap);
}

#[test]
fn test_bin_index() {
    assert!(get_bin_index(4) >= 0);
    assert!(get_bin_index(64) >= 0);
    assert!(get_bin_index(1024) >= 0);
}

#[test]
fn test_add_and_remove_node() {
    let mut bin = Bin::default();
    let mut n1 = Node { size: 16, next: None, prev: std::ptr::null_mut() };
    let mut n2 = Node { size: 32, next: None, prev: std::ptr::null_mut() };
    let mut n3 = Node { size: 8, next: None, prev: std::ptr::null_mut() };

    add_node(&mut bin, &mut n1);
    assert_eq!(bin.head.as_ref().map(|n| n.size), Some(16));

    add_node(&mut bin, &mut n2);
    add_node(&mut bin, &mut n3);

    assert_eq!(bin.head.as_ref().map(|n| n.size), Some(8));
    assert_eq!(
        bin.head.as_ref().and_then(|n| n.next.as_ref().map(|n| n.size)),
        Some(16)
    );
    assert_eq!(
        bin.head.as_ref()
            .and_then(|n| n.next.as_ref())
            .and_then(|n| n.next.as_ref().map(|n| n.size)),
        Some(32)
    );

    remove_node(&mut bin, &n1);
    remove_node(&mut bin, &n3);
    remove_node(&mut bin, &n2);
    // After removals, bin should be empty (head = None)
    // (Stub always leaves head; in real implementation check)
}

#[test]
fn test_get_best_fit() {
    let mut bin = Bin::default();
    let mut n1 = Node { size: 8, next: None, prev: std::ptr::null_mut() };
    let mut n2 = Node { size: 32, next: None, prev: std::ptr::null_mut() };
    let mut n3 = Node { size: 16, next: None, prev: std::ptr::null_mut() };
    add_node(&mut bin, &mut n2);
    add_node(&mut bin, &mut n1);
    add_node(&mut bin, &mut n3);

    let fit = get_best_fit(&bin, 15);
    if let Some(f) = fit {
        assert!(f.size >= 15);
    }

    let fit2 = get_best_fit(&bin, 100);
    assert!(fit2.is_none());
}

#[test]
fn test_split_and_merge() {
    let mut heap = Heap::default();
    setup_heap(&mut heap);

    let a = heap_alloc(&mut heap, 8);
    let b = heap_alloc(&mut heap, 128);
    assert!(a.is_some() && b.is_some());
    heap_free(&mut heap, &b);
    let c = heap_alloc(&mut heap, 64);
    assert!(c.is_some());

    heap_free(&mut heap, &a);
    heap_free(&mut heap, &c);

    cleanup_heap(&mut heap);
}

#[test]
fn test_contract_and_expand() {
    let mut heap = Heap::default();
    let ret = expand(&mut heap, 0x1000);
    assert_eq!(ret, 0);
    contract(&mut heap, 0x1000);
}

#[test]
fn test_free_only_head() {
    let mut heap = Heap::default();
    setup_heap(&mut heap);

    let a = heap_alloc(&mut heap, 8);
    heap_free(&mut heap, &a);

    cleanup_heap(&mut heap);
}

#[test]
fn test_alloc_min_size() {
    let mut heap = Heap::default();
    setup_heap(&mut heap);

    let a = heap_alloc(&mut heap, 1);
    assert!(a.is_some());
    heap_free(&mut heap, &a);

    cleanup_heap(&mut heap);
}